### PR TITLE
Add recipe for libmetatensor-torch

### PR DESCRIPTION
**Pull request by Luthaf**
This is the continuation of https://github.com/conda-forge/metatensor-feedstock/issues/2, now packaging the second native library. This one is pure C++, and links to libtorch.

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by Luthaf** [comment]
Just to check, the version of libtorch is pinned across the ecosystem, right? And any bump in the libtorch version will trigger a rebuild of all feedstocks depending on it?

**Comment by Luthaf** [comment]
@conda-forge/help-c-cpp ready for review!

File: recipes/libmetatensor-torch/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+test:
+  requires:
+    - cmake
+    - ninja
+    - {{ compiler('cxx') }}
+    - libtorch  =*=cpu*
+  files:
+    - test-project/
+  commands:
```
Comment (xhochy) [inline recipes/libmetatensor-torch/meta.yaml:43]: Please add a simple file existence test for checking that the shared library is part of the package and at the correct path.

File: recipes/libmetatensor-torch/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+package:
+  name: libmetatensor-torch
+  version: {{ version }}
+
+source:
+  url: {{ url_base }}/metatensor-torch-v{{ version }}/metatensor-torch-cxx-{{ version }}.tar.gz
+  sha256: 904cf858d8f98b67b948e8a453d8a6da56111e022050d6c8c3d32a9a2cc83464
+
+build:
```
Comment (xhochy) [inline recipes/libmetatensor-torch/meta.yaml:12]: Shared libraries also need a run_exports.

**Review by xhochy — CHANGES_REQUESTED** [review]

**Comment by PicoCentauri** [comment]
Thanks for the work @Luthaf. Of course I am willing to help maintaining this recipe.

File: recipes/libmetatensor-torch/meta.yaml (hunk 2 above)
Comment (Luthaf) [reply recipes/libmetatensor-torch/meta.yaml:12]: It's not very clear what run_exports does. Does it only add a `run` dependency to all packages that built against the library and had it as a `host` dependency? Or is there something else I need to be aware of here?

Comment (carterbox) [reply recipes/libmetatensor-torch/meta.yaml:12]: https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#export-runtime-requirements

run_exports export runtime requirements to downstream packages which dynamically link to your package. It's a statement which says "hey, if you link to my package, then this is the version that you will need at runtime."

Comment (Luthaf) [reply recipes/libmetatensor-torch/meta.yaml:12]: Thanks! I've added `run_exports` to [`libmetatensor`](https://github.com/conda-forge/libmetatensor-feedstock) and removed it from the `run` requirements here. 

I'd guess `libtorch` also has a run_exports, but is there a way to express "build against the cpu variant but run with any variant, cpu or cuda"? For now I do two separate host & run requirements.

**Comment by Luthaf** [comment]
@conda-forge/help-c-cpp ready for another round of review!

File: recipes/libmetatensor-torch/meta.yaml (hunk 2 above)
Comment (carterbox) [reply recipes/libmetatensor-torch/meta.yaml:12]: > I'd guess libtorch also has a run_exports

[Yes.
](https://github.com/conda-forge/pytorch-cpu-feedstock/blob/2953d76cb7cfa5fd5b8049c1df9baf0758bab4af/recipe/meta.yaml#L43)

Looks like the run_exports for `libtorch` are variant agnostic, so if you build against any variant of `libtorch` you can get any variant at runtime. It looks like this package doesn't link against `libtorch_cuda`, so you don't need to worry about contraining the `libtorch` build variant.

File: recipes/libmetatensor-torch/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    - {{ stdlib("c") }}
+    - cmake
+    - ninja
+  host:
+    - libmetatensor >=0.1.8,<0.2
+    # we only need the CPU version of libtorch to build the code, but we can run
+    # with either the CPU or CUDA version
+    - libtorch  =*=cpu*
+  run:
+    - libtorch
```
Comment (carterbox) [inline recipes/libmetatensor-torch/meta.yaml:35]:
```suggestion
    # We don't link to any gpu libraries
    - libtorch
```

**Comment by carterbox** [comment]
The logs show that the build system is also looking for zlib? Is that a missing host dependency?

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=935468&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=808

**Comment by Luthaf** [comment]
> The logs show that the build system is also looking for zlib? Is that a missing host dependency?

We are not trying to load ZLIB ourself. Digging around the CMake files, this is what seems to happen:

We load torch with `find_package(Torch)`, which then does `find_package(Caffe2)`

https://github.com/pytorch/pytorch/blob/6c503f1dbbf9ef1bf99f19f0048c287f419df600/cmake/TorchConfig.cmake.in#L68

Which then does `find_package(Protobuf)`: 

https://github.com/pytorch/pytorch/blob/6c503f1dbbf9ef1bf99f19f0048c287f419df600/cmake/public/protobuf.cmake#L4

which finally tries to find ZLIB:

https://github.com/protocolbuffers/protobuf/blob/2f3242c576504459621fd7d78bbccf39bfaa49c5/cmake/protobuf-config.cmake.in#L5
https://github.com/protocolbuffers/protobuf/blob/2f3242c576504459621fd7d78bbccf39bfaa49c5/CMakeLists.txt#L172

**Pull request merged**